### PR TITLE
Stabilize the `re_rebalance_coherence` feature

### LINKED ISSUE #63599: Stabilize RFC 2451, re-rebalance coherence
# Stabilization Proposal

I am writing to propose we stabilize the implementation of [RFC 2451], "re-rebalancing coherence" ([tracking issue]).

[RFC 2451]: https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html
[tracking issue]: https://github.com/rust-lang/rust/issues/55437

## Brief Summary

The RFC modifies the coherence rules. It used to be that, when implementing a non-local trait, if Ti was the first "local type", then all type parameters were forbidden before that point. In practice, that meant that this impl was illegal:

```rust
struct Local;

impl<T> Foreign<Local> for Vec<T> { }
//              ^^^^^          ^
//              |              type parameter
//              first local type
```

The RFC lightly generalizes these rules. In particular, type parameters are now allowed before Ti, but *only* if they are **covered**, defined in the RFC as:

> Covered Type: A type which appears as a parameter to another type. For example, `T` is uncovered, but the `T` in `Vec<T>` is covered. This is only relevant for type parameters.

This means that the impl above is now legal, because the type`T` is covered by `Vec`. Note that an impl like the following would still be illegal:

```rust
struct Local;

impl<T> Foreign<Local> for T { }
//              ^^^^^      ^
//              |          type parameter (uncovered)
//              first local type
```

It is considered a (semver) breaking change for parent crates to add impls with uncovered type parameters (this was already the case even with the old rules, I believe). Therefore, this generalization of the rules cannot permit a conflict between the parent crate and a child crate. ([See RFC for a more detailed argument.](https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html#concrete-orphan-rules))

(NB: I've elided details of `#[fundamental]` in this summary, see the RFC for full details.)

## Changes since the RFC was approved

None

## Test cases

This section describes the patterns covered by new tests added for this RFC, or tests that seemed relevant to this RFC, with links to the test files.

- `impl<T> Foreign<Local> for Foreign<T, Foreign>` -- [re-rebalance-coherence.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/re-rebalance-coherence.rs#L11), [re-rebalance-coherence-rpass.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/coherence/re-rebalance-coherence-rpass.rs#L12) (these appear to be duplicate tests)
- `impl<T> Foreign for Foreign<T, Local<T>>` -- [coherence-cow.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-cow.rs#L18), [coherence-cow.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-cow.rs#L23)
- `impl<T,U> Foreign for Foreign<T, Local<U>>` -- [coherence-cow.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-cow.rs#L28), [coherence-pair-covered-uncovered.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-lone-type-parameter.rs#L9)
- `impl<T> Foreign<Local> for T` -- [coherence-bigint-param.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-bigint-param.rs#L11)
- `impl Foreign for Local<T>` -- [coherence-covered-type-parameter.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-covered-type-parameter.rs#L15), [coherence-iterator-vec.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-iterator-vec.rs#L15)
- `impl<T> Foreign<T> for Foreign` -- [coherence-all-remote.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-all-remote.rs#L9)
- `impl Foreign<Local> for Foreign` -- [coherence-bigint-int.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-bigint-int.rs#L14), [coherence-bigint-vecint.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-bigint-vecint.rs#L14)
- `impl<T> Foreign for T` -- [coherence-cross-crate-conflict.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-cross-crate-conflict.rs#L12), [coherence-lone-type-parameter.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-lone-type-parameter.rs#L9)
- `impl Foreign for Foreign<Local>` -- [coherence-vec-local.rs](https://github.com/rust-lang/rust/blob/1cdcea920e56a5d0587307a4c9cf8fff5c77c4bc/src/test/ui/coherence/coherence-vec-local.rs#L14)
- `impl<T> Foreign<Foreign<T>, Local> for Foreign<T>` -- in https://github.com/rust-lang/rust/pull/64414
- `impl<T> Foreign<Local> for Fundamental<T>` -- in https://github.com/rust-lang/rust/pull/64414

## Missing items

- [ ] Open PR against the reference, updating documentation
- [x] Fix https://github.com/rust-lang/rust/issues/64412
- [ ] Add more tests for fundamental patterns -- we did not by any means tile the space

### PR DESCRIPTION
**Pull request by ohadravid**
This PR stabilizes [RFC 2451](https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html), re-rebalance coherence.

Changes include removing the attribute from tests which tested both the old and new behavior, moving the feature to `accepted` and removing the old logic.

I'll also open a [PR](https://github.com/rust-lang-nursery/reference/pull/703) against the reference, updating it with the content of the RFC.

Closes #63599

r? @nikomatsakis

### CODE REVIEW

**Comment by ohadravid** [comment]
Apparently the example used in the [documentation](https://github.com/rust-lang/rust/blob/master/src/libcore/convert.rs#L227) for the `Into` trait fails because it can now be complied.

This is the expected behavior (and one of the RFC's stated goals), but I don't know what to do with it (remove the example? change it to say that `before rust. 1.40...`?).

**Comment by nikomatsakis** [comment]
@ohadravid 

> Apparently the example used in the documentation for the Into trait fails because it can now be complied.

amusing

**Comment by nikomatsakis** [comment]
I wonder if this means that whole section can be removed, and we can just advice to *always* implement `From`?

File: src/libcore/convert.rs [hunk 1]
```diff
@@ -220,21 +220,20 @@ pub trait AsMut<T: ?Sized> {
 ///
 /// # Implementing [`Into`] for conversions to external types
 ///
-/// If the destination type is not part of the current crate
-/// then you can't implement [`From`] directly.
+/// Prior to Rust 1.40, if the destination type was not part of the current crate
```
Comment (nikomatsakis) [inline src/libcore/convert.rs]: I think we should strengthen this -- unless we have a good example of when you *do* need to implement `Into`, we should probably start by telling people to **avoid** implementing `Into` directly, and then continue with this content?

**Review by nikomatsakis — APPROVED** [review]
This looks great! I left one suggestion for how to approach the docs -- I think we could be a bit more forceful -- but I'm also relatively happy as is.

**Comment by ohadravid** [comment]
As far as I can tell, there is no reason to implement `Into`, so I changed it to be more forceful.
However I just might not be able to come up with a contrived-enough example.

(I went to `diesel` and changed all the `Into` to `From` and it worked where it didn't before `re_rebalance`, so that's something)

**Comment by ohadravid** [comment]
@nikomatsakis I changed the docs like you suggested! Let me know if there's anything else to do 😁

File: src/libsyntax/feature_gate/accepted.rs [hunk 2]
```diff
@@ -253,6 +253,9 @@ declare_features! (
     (accepted, const_constructor, "1.40.0", Some(61456), None),
     /// Allows the use of `#[cfg(doctest)]`, set when rustdoc is collecting doctests.
     (accepted, cfg_doctest, "1.40.0", Some(62210), None),
+    /// Allows relaxing the coherence rules such that
+    /// `impl<T> ForeignTrait<LocalType> for ForeignType<T>` is permitted.
+    (accepted, re_rebalance_coherence, "1.40.0", Some(55437), None),
```
Comment (Centril) [inline src/libsyntax/feature_gate/accepted.rs]:
```suggestion
    (accepted, re_rebalance_coherence, "1.41.0", Some(55437), None),
```

**Review by nikomatsakis — APPROVED** [review]

File: src/libcore/convert.rs (hunk 1 above)
Comment (dalcde) [reply src/libcore/convert.rs]: This should say 1.41 instead?

**Pull request merged**